### PR TITLE
KAFKA-14443: Close topic creation Admin clients in MM2 connectors

### PR DESCRIPTION
**Pull request by gharris1727**
The MirrorCheckpointConnector, MirrorHeartbeatConnector, and MirrorSourceConnector all create internal topics, but leak the Admin client used to do so. Instead, this Admin client should be persisted with the lifetime of the connector, and closed when the connector is stopped.

Additionally, generalize some types from ForwardingAdmin to Admin where the subtype is not a requirement.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by gharris1727** [comment]
cc @OmniaGM @tombentley @mimaison

This is a regression introduced in #12577 and should be merged to all branches which have that feature (AFAIU, only trunk presently.)

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnector.java
Comment (OmniaGM) [inline connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnector.java:169]: Why are we replacing `config.forwardingAdmin(config.targetAdminConfig())` which creates the target admin client with the source admin client here?

Comment (OmniaGM) [reply connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnector.java:169]: You also might need to close the target admin client in this connector as well.

**Review by OmniaGM — CHANGES_REQUESTED** [review]
Thanks for the pr, there are no other branches related to KIP-787 which was implemented in #12577. I just left a couple of comments on this pr.

**Review by OmniaGM — COMMENTED** [review]
Thanks, @gharris1727 for the updates. You have one integration test failing here https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12955/2/testReport/org.apache.kafka.connect.mirror.integration/ Other than this the pr LGTM.

**Review by C0urante — COMMENTED** [review]
Thanks @gharris1727. I've verified that every `Admin` created via `MirrorConnectorConfig::forwardingAdmin` is now closed correctly, and that the single one created with `MirrorClient::forwardingAdmin` was already being closed correctly.

I'd like to see some testing coverage but I'm honestly not sure how we might do that in a way that's minimally invasive to the non-testing portions of the code base, and provides useful guarantees that this same bug won't come up again in the future.

It's not enough to just verify that all admin clients allocated in `start` are also closed in `stop`, since one of the root causes of this bug is that `MirrorConnectorConfig::forwardingAdmin` can be invoked at any time throughout a connector or a task's lifecycle.

One possible approach--that may also make it harder to write bugs like this in the first place without having to run tests to catch them--could be to shift the responsibility for instantiating and closing the admin clients to the various `Mirror*Config` classes, which can expose a method like `close` (or, if that's too generic, `releaseAdminClients`). We could then verify that this method is invoked in the `stop` method of each relevant connector and task class, reduce the visibility of `MirrorConnectorConfig::forwardingAdmin` so that it can't be invoked directly by connector/task classes, and add testing for the config classes to ensure that they create and close the expected clients.

Thoughts?

**Comment by gharris1727** [comment]
@OmniaGM I believe that failure is flaky, as I cannot reproduce it locally. Those sorts of flaky failures are what got me looking at this code in the first place, coincidentally.

@C0urante Thanks for your suggestion on making these sorts of leaks more difficult, given the complexity and number of resources being managed by the connector/task instances, that we should certainly look into it.

I thought about the `Mirror*Config::close` solution that you proposed, and didn't like the idea of the config container being also responsible for managing the lifetime of the objects it creates. I didn't like it being responsible for creating the resources either, as you pointed out: it's too easy to leak the reference.

I experimented with an alternative solution, to have a reusable (across task/connector classes) BackgroundResources class which handles all instantiation and cleanup of the AutoClosable resources used in an MM2 connector/task. It simplifies the `stop()` methods, and doesn't appear to incur too much boilerplate in the `start()` methods, maybe with a shorter name it could be even better.

I have two concerns here:
1. This change may be scope-creep for this PR, and maybe better handled as a lateral refactor after the release blocker bug is resolved. I am happy to continue iterating on the wider solution in this PR or a separate one, so let me know if you think the scope is still reasonable.
2. I wonder if the BackgroundResources class is in danger of becoming a "god object" which does everything. At the moment everything passes checkstyle, but if we add too many other pluggable interfaces or closeable resources, this class could be doing too much.

**Comment by ableegoldman** [comment]
> This change may be scope-creep for this PR, and maybe better handled as a lateral refactor after the release blocker bug is resolved.

Just my 2 cents as the 3.4 release manager -- unless this refactor helps to stabilize the actual feature as-is or fixes a suspected issue, please save this for a followup PR and avoid increasing the scope here: both to reduce the possibility of introducing new bugs with a nontrivial last-minute refactor, and of course to help ensure this fix is merged in a timely manner.

Happy to discuss further especially since I have absolutely zero context on the particular issue or this part of the codebase in general, but just keep in mind how close we are to the end of the 3.4 release. Thanks 🙂

**Comment by C0urante** [comment]
@ableegoldman Thanks for taking a look--I agree that a more thorough refactoring can wait for a follow-up PR.

@gharris1727 I really like the general approach in the new `BackgroundResources` class and would love to still try to get something like it merged, but agree with your (and Sophie's) concerns about scope creep, especially with the upcoming release.

Given that adding any sort of meaningful tests for a fix to this issue would require a non-negligible refactoring, and the simplicity of the original fix proposed here (which simply added a few instance variables to a few classes, all of which were closed later), I think we can revert this PR to its original approach and file a follow-up PR for more sophisticated resource-tracking later. We can plan to merge it to trunk only after the 3.4 branch has been cut.

**Review by C0urante — APPROVED** [review]
Thanks Greg! LGTM; will merge pending CI build.

**Comment by C0urante** [comment]
The integration test failures, though worrying, do not appear to be caused by these changes and have been occurring for some time now. I've filed https://issues.apache.org/jira/browse/KAFKA-14453 to note that they're flaky, and feel comfortable merging this as-is due to the extremely low risk of regression.

**Comment by C0urante** [comment]
Merged and backported to the 3.4 release branch.

**Comment by ableegoldman** [comment]
Thanks guys, appreciate you splitting out the refactoring even though it's a hassle (and maybe doesn't feel as good) 🙏

**Pull request merged**
